### PR TITLE
feat(core-flows,medusa,types,utils): adds update cart API with promotions

### PR DESCRIPTION
**Pull request by riqwan**
what:

- adds update cart API
  - workflow
  - promotions
  - sales channel
  - region
  - customer

### CODE REVIEW

File: integration-tests/plugins/__tests__/cart/store/carts.spec.ts [hunk 1]
```diff
@@ -288,6 +292,214 @@ describe("Store Carts API", () => {
...
+            expect.objectContaining({
+              id: "item-1",
+              adjustments: [],
+            }),
+          ],
+        })
+      )
+    })
+
+    it("should update a cart's region", async () => {
```
Comment (olivermrbl) [inline integration-tests/plugins/__tests__/cart/store/carts.spec.ts]: suggestion(non-blocking): Squash the region, customer, and sales channel tests.

File: packages/core-flows/src/definition/cart/steps/update-carts.ts [hunk 2]
```diff
@@ -1,61 +1,44 @@
...
-  selector: FilterableCartProps
-  update: UpdateCartDataDTO
-}
-
 export const updateCartsStepId = "update-carts"
 export const updateCartsStep = createStep(
   updateCartsStepId,
-  async (data: UpdateCartsStepInput, { container }) => {
-    const service = container.resolve<ICartModuleService>(
+  async (data: CartWorkflow.UpdateCartWorkflowInputDTO, { container }) => {
```
Comment (olivermrbl) [inline packages/core-flows/src/definition/cart/steps/update-carts.ts]: thought(non-blocking): Can we keep this as a bulk step? Unless there is a specific reason for converting it to a single cart update, I'd like that.

File: packages/core-flows/src/definition/cart/workflows/update-carts.ts [hunk 3]
```diff
@@ -1,20 +1,114 @@
...
+  findSalesChannelStep,
+  getActionsToComputeFromPromotionsStep,
+  prepareAdjustmentsFromPromotionActionsStep,
+  removeLineItemAdjustmentsStep,
+  removeShippingMethodAdjustmentsStep,
+  retrieveCartStep,
+  updateCartsStep,
+} from "../steps"
 
 export const updateCartsWorkflowId = "update-carts"
```
Comment (olivermrbl) [inline packages/core-flows/src/definition/cart/workflows/update-carts.ts]: nit(non-blocking): Rename to `updateCartWorkflow` -> operation on a single cart

File: packages/medusa/src/api-v2/store/carts/validators.ts [hunk 4]
```diff
@@ -81,14 +81,19 @@ export class StorePostCartsCartReq {
...
   sales_channel_id?: string
 
   @IsObject()
   @IsOptional()
   metadata?: Record<string, unknown>
 
+  @IsOptional()
+  @IsArray()
+  @Type(() => String)
+  promo_codes: string[]
```
Comment (olivermrbl) [inline packages/medusa/src/api-v2/store/carts/validators.ts]: nit:
```suggestion
  promo_codes?: string[]
```

File: packages/types/src/workflow/cart/update-cart.ts [hunk 5]
```diff
@@ -0,0 +1,10 @@
+export interface UpdateCartWorkflowInputDTO {
```
Comment (olivermrbl) [inline packages/types/src/workflow/cart/update-cart.ts]: nit: I believe I introduced cart workflow types in `types/src/cart/workflows.ts`. Perhaps, we can consolidate. Nothing important, so consider this non-blocking

File: packages/core-flows/src/definition/cart/workflows/update-carts.ts [hunk 6]
```diff
@@ -1,20 +1,114 @@
+import { CartDTO, CartWorkflow } from "@medusajs/types"
```
Comment (olivermrbl) [inline packages/core-flows/src/definition/cart/workflows/update-carts.ts]: nit(non-blocking): Rename file `update-cart.ts` -> operation on a single cart

File: packages/types/src/cart/workflows.ts [hunk 7]
```diff
@@ -65,3 +65,14 @@ export interface AddToCartWorkflowInputDTO {
   items: CreateCartCreateLineItemDTO[]
   cart: CartDTO
 }
+
+export interface UpdateCartWorkflowInputDTO {
+  id: string
+  promo_codes?: string[]
+  region_id?: string
```
Comment (sradevski) [inline packages/types/src/cart/workflows.ts:72]: Can you even update the region of a cart? I am still exploring what is a possibility or not so bear with me :D

The same question on customer_id, sales_channel_id and currency_code I guess.

Comment (riqwan) [reply packages/types/src/cart/workflows.ts:72]: Good question 😄 I'm not 100% sure either.

From the customer_id perspective, I think its possible that you'd want to move an anonymous customer to a customer thats logged in mid cart. I suppose a customer could also update region/currency/sales channels, but that should have some consequences on the items that are already present in the cart, which is not done in this PR. 

Why these things are done are entirely dependent on the consumer's usecases and workflows.

Comment (sradevski) [reply packages/types/src/cart/workflows.ts:72]: @riqwan I think adding some rail guards is still part of the module's responsibility, so we should either not allow eg. modifying a region, or ensure that if one does do that, we bring the cart into a consistent state (updating what items are in the cart, etc.). I guess we could shift the responsibility to a workflow, but this seems like a core responsibility of the cart.

Anyway, these are discussions that can happen later on as well, so it's not a blocker for sure, it's all about where responsibilities lie ultimately.

Comment (riqwan) [reply packages/types/src/cart/workflows.ts:72]: Yup, completely agree there.

**Pull request merged**